File: SourceCode/backend/app/schemas/analysis.py
from typing import Literal

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel


class CamelModel(BaseModel):
    """Base for schemas that must serialize as camelCase to match the frontend's
    types (src/types/analysis.ts) exactly."""

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class WorkflowStep(CamelModel):
    n: int
    time: str
    title: str
    description: str


class ClassificationResult(CamelModel):
    """Matches Eshwaran's real /jobs/{id}/results shape (label/confidence/
    probabilities) agreed on 2026-09-05 -- this is the classifier's raw
    output, not a narrative. See SopReport for the separate SOP writeup."""

    id: str
    name: str
    date: str
    duration: str
    status: Literal["Complete"]
    video_url: str | None = None
    label: str
    confidence: float
    probabilities: dict[str, float]


class SopReport(CamelModel):
    """The narrative writeup (summary + step-by-step breakdown), fetched
    separately via GET /jobs/{id}/sop. Prototype data until the real
    LLM/RAG generation stage exists (ReadMe.md timeline: Oct 15)."""

    summary: str
    steps: list[WorkflowStep]


class VideoUploadResponse(BaseModel):
    id: str


class JobStatusResponse(CamelModel):
    status: Literal["processing", "complete", "failed"]
    progress: float
    error: str | None = None

File: SourceCode/frontend/src/types/analysis.ts
export interface WorkflowStep {
  n: number;
  time: string;
  title: string;
  description: string;
}

/** Matches the real backend's /jobs/{id}/results shape (agreed with Eshwaran
 * 2026-09-05) -- the classifier's raw output, not a narrative. */
export interface ClassificationResult {
  id: string;
  name: string;
  date: string;
  duration: string;
  status: 'Complete';
  videoUrl: string | null;
  /** One of the 10 classes in the project ReadMe.md's "Target Classes" list. */
  label: string;
  confidence: number;
  probabilities: Record<string, number>;
}

/** The narrative SOP writeup, fetched separately from the classification.
 * Prototype data until the real LLM/RAG generation stage exists. */
export interface SopReport {
  summary: string;
  steps: WorkflowStep[];
}

export interface SearchResultItem {
  jobId: string;
  videoName: string;
  snippet: string;
  score: number;
}

export type Screen = 'upload' | 'analyzing' | 'results' | 'history' | 'dashboard' | 'search' | 'report';
